libraries/AP_ExternalAHRS/AP_ExternalAHRS_GSOF.cpp

```diff
@@ -0,0 +1,362 @@
+//     $ sim_vehicle.py -v Plane -A "--serial3=uart:/dev/ttyUSB0" --console --map -DG
+//     param set SERIAL3_PROTOCOL 36
+//     param set SERIAL3_BAUD 115200
+//     param set EAHRS_TYPE 6
+//     # ensure NET_* if off if you were using ethernet.
+//     # To enable the EAHRS to provide GPS:
+//     param set GPS2_TYPE 21
+
+// On most hardware, you must enable EAHRS:
+//     ./waf configure --board Pixhawk6X --enable-AHRS_EXT
+
+// GPS ride-along with EARHS as the 2nd GPS
+//      param set GPS_AUTO_SWITCH 0
+//      param set GPS2_TYPE 21
+//      param set GPS_PRIMARY 0
+
+#define AP_MATH_ALLOW_DOUBLE_FUNCTIONS 1
+
+#include "AP_ExternalAHRS_config.h"
+
+#if AP_EXTERNAL_AHRS_GSOF_ENABLED
+
+#include "AP_ExternalAHRS_GSOF.h"
+#include "AP_Compass/AP_Compass_config.h"
+#include <AP_Baro/AP_Baro.h>
+#include <AP_Compass/AP_Compass.h>
+#include <AP_GPS/AP_GPS.h>
+#include <AP_HAL/utility/sparse-endian.h>
+#include <AP_InertialSensor/AP_InertialSensor.h>
+#include <GCS_MAVLink/GCS.h>
+#include <AP_Logger/AP_Logger.h>
+#include <AP_BoardConfig/AP_BoardConfig.h>
+#include <AP_HAL/utility/Socket.h>
+
+static const char* LOG_FMT = "%s ExternalAHRS: %s";
+
+extern const AP_HAL::HAL &hal;
+
+AP_ExternalAHRS_GSOF::AP_ExternalAHRS_GSOF(AP_ExternalAHRS *_frontend,
+        AP_ExternalAHRS::state_t &_state): AP_ExternalAHRS_backend(_frontend, _state)
+{
+
+    auto &sm = AP::serialmanager();
+    uart = sm.find_serial(AP_SerialManager::SerialProtocol_AHRS, 0);
+    if (!uart) {
+        GCS_SEND_TEXT(MAV_SEVERITY_ERROR, LOG_FMT, get_name(), "no UART");
+        return;
+    }
+    baudrate = sm.find_baudrate(AP_SerialManager::SerialProtocol_AHRS, 0);
+    port_num = sm.find_portnum(AP_SerialManager::SerialProtocol_AHRS, 0);
+    pktbuf = NEW_NOTHROW uint8_t[AP_GSOF::MAX_PACKET_SIZE];
+
+    if (!pktbuf) {
+        AP_BoardConfig::allocation_error("GSOF ExternalAHRS");
+    }
+
+    if (!hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_ExternalAHRS_GSOF::update_thread, void), "AHRS", 2048, AP_HAL::Scheduler::PRIORITY_SPI, 0)) {
+        AP_BoardConfig::allocation_error("GSOF ExternalAHRS failed to allocate ExternalAHRS update thread");
+    }
+
+    // Offer GPS even through it's a tightly coupled EKF.
+    set_default_sensors(uint16_t(AP_ExternalAHRS::AvailableSensor::GPS));
+
+    hal.scheduler->delay(5000);
+    if (!initialised()) {
+        GCS_SEND_TEXT(MAV_SEVERITY_WARNING, LOG_FMT, get_name(), "missing data within 5s.");
+    }
+}
+
+// get serial port number for the uart
+int8_t AP_ExternalAHRS_GSOF::get_port(void) const
+{
+    if (!uart) {
+        return -1;
+    }
+    return port_num;
+};
+
+void AP_ExternalAHRS_GSOF::update_thread(void)
+{
+    // TODO configure receiver to output expected data.
+
+    auto last_debug = AP_HAL::millis();
+    size_t pps = 0;
+
+    uart->begin(baudrate);
+    
+    while (true) {
+        // TODO should we ever call begin() again?
+        while (uart->available() > 0) {
```

Comment:
You are going to end up tight-looping in this thread at a ridiculously high priority while data isn't available on the UART.

Probably need a short sleep if there's no data to be read, meaning a change to this:

```suggestion
        const auto available = uart->available();
        if (available == 0) {
           .... sleep
            continue;
        }
        for (auto i=0; i<available; i++) {
```

Reply:
Good point, the other approach Tridge suggested would not have solved this case.
Fixed in 0e3355a5991cf630ebe29c89e463e4224b3855d9